src/java/org/apache/cassandra/cql3/restrictions/MergedRestriction.java

```diff
@@ -0,0 +1,334 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.restrictions;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.RangeSet;
+
+import org.apache.cassandra.cql3.Operator;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.functions.Function;
+import org.apache.cassandra.db.filter.RowFilter;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.IndexRegistry;
+import org.apache.cassandra.schema.ColumnMetadata;
+
+import static org.apache.cassandra.cql3.statements.RequestValidations.invalidRequest;
+
+/**
+ * A {@code SingleRestriction} which is the result of merging multiple {@code SimpleRestriction}.
+ */
+public final class MergedRestriction implements SingleRestriction
+{
+    /**
+     * The columns to which the restrictions apply.
+     */
+    private final List<ColumnMetadata> columns;
+
+    /**
+     * The restrictions composing this restriction
+     */
+    private final List<SimpleRestriction> restrictions;
+
+    private final boolean isOnToken;
+
+    private final boolean isSlice;
+
+    private final boolean isMultiColumn;
+
+    /**
+     * The number of restrictions that require {@code CONTAINS}, {@code CONTAINS_KEY} and Map equality restrictions.
+     */
+    private final int containsCount;
+
+    public MergedRestriction(SingleRestriction restriction,
+                             SimpleRestriction other)
+    {
+        assert restriction.isOnToken() == other.isOnToken();
+
+        this.columns = restriction.columns().size() < other.columns().size()
```

Comment:
Are you assuming here one to be a superset of the other? both to be based off a common base set of columns? If so the javadoc should mention it imo

Reply:
Yes it is the assumption.